File: javascript/src/solution/Item.js
const FactoryItem = (() => {

    const nameKey = Symbol('name');
    const sellInKey = Symbol('sellIn');
    const qualityKey = Symbol('sellIn');
    const updateSellInKey = Symbol('updateSellIn()');
    const updateQualityKey = Symbol('updateQuality()');

    return class Item{
        constructor(name, sellIn, quality){
            this[nameKey] = name;
            this[sellInKey] = sellIn;
            this[qualityKey] = quality;

            this[updateSellInKey] =  function updateSellIn(){
                this[sellInKey] =  this[sellInKey] -1;
            };

            this[updateQualityKey] =  function updateQuality(){
                if( this[qualityKey] > 0){
                    if(this[sellInKey] <= 0){
                        this[qualityKey] =  this[qualityKey] -2;
                    }else{ this[qualityKey] =  this[qualityKey] -1;}
                }
            };
        }

        updateItem(){
            this[updateSellInKey]();
            this[updateQualityKey]();
        };

        print(){
            return this[nameKey] + ", " +  this[sellInKey] + ", " +  this[qualityKey];
        };

        getName(){
         return this[nameKey];
        };

        getSellIn(){
            return  this[sellInKey];
        };

        setSellIn(newSellIn){
            this[sellInKey] = newSellIn;
        };

        getQuality(){
            return  this[qualityKey];
        };

        setQuality(newQuality){
            this[qualityKey] = newQuality;
        };
    }
});

module.exports = FactoryItem();

File: java/issue/src/solution/Item.java
package solution;

public class Item {

    private String name;

    protected int sellIn;

    protected int quality;

    Item(String name, int sellIn, int quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }

    void updateItem(){
        updateSellIn();
        updateQuality();
    }

    protected void updateSellIn() {
        sellIn = sellIn -1;
    }

    protected void updateQuality() {
        if(quality > 0){
            if(sellIn <= 0){
                quality = quality -2;
            }else{
                quality = quality -1;
            }
        }
    }

    @Override
    public String toString() {
        return this.name + ", " + this.sellIn + ", " + this.quality;
    }

    String getName() {
        return name;
    }

    int getSellIn() {
        return sellIn;
    }

    int getQuality() {
        return quality;
    }

}